src/java.base/share/classes/java/util/ReverseOrderSortedMapView.java

```diff
@@ -0,0 +1,464 @@
+        return base.containsKey(key);
+    }
+
+    public boolean containsValue(Object value) {
+        return base.containsValue(value);
+    }
+
+    public V get(Object key) {
+        return base.get(key);
+    }
+
+    public boolean isEmpty() {
+        return base.isEmpty();
+    }
+
+    public V put(K key, V value) {
+        return base.put(key, value);
+    }
+
+    public void putAll(Map<? extends K, ? extends V> m) {
+        base.putAll(m);
+    }
+
+    public V remove(Object key) {
+        return base.remove(key);
+    }
+
+    public int size() {
+        return base.size();
+    }
+
+    public Set<K> keySet() {
+        return new AbstractSet<>() {
+            // inherit add(), which throws UOE
+            public Iterator<K> iterator() { return descendingKeyIterator(base); }
+            public int size() { return base.size(); }
+        };
+    }
+
+    public Collection<V> values() {
+        return new AbstractCollection<>() {
+            // inherit add(), which throws UOE
+            public Iterator<V> iterator() { return descendingValueIterator(base); }
+            public int size() { return base.size(); }
+        };
+    }
+
+    public Set<Entry<K, V>> entrySet() {
+        return new AbstractSet<>() {
+            // inherit add(), which throws UOE
+            public Iterator<Entry<K, V>> iterator() { return descendingEntryIterator(base); }
+            public int size() { return base.size(); }
+        };
+    }
+
+    // ========== SequencedMap ==========
+
+    public SortedMap<K, V> reversed() {
+        return base;
+    }
+
+    public K firstKey() {
+        return base.lastKey();
+    }
+
+    public K lastKey() {
+        return base.firstKey();
+    }
+
+    public Map.Entry<K, V> firstEntry() {
+        return base.lastEntry();
+    }
+
+    public Map.Entry<K, V> lastEntry() {
+        return base.firstEntry();
+    }
+
+    public Map.Entry<K,V> pollFirstEntry() {
+        return base.pollLastEntry();
+    }
+
+    public Map.Entry<K,V> pollLastEntry() {
+        return base.pollFirstEntry();
+    }
+
+    public V putFirst(K k, V v) {
+        return base.putLast(k, v);
+    }
+
+    public V putLast(K k, V v) {
+        return base.putFirst(k, v);
+    }
+
+    // ========== SortedMap ==========
+
+    public Comparator<? super K> comparator() {
+        return cmp;
+    }
+
+    public SortedMap<K, V> subMap(K fromKey, K toKey) {
+        if (cmp.compare(fromKey, toKey) <= 0) {
+            return new Submap(fromKey, toKey);
+        } else {
+            throw new IllegalArgumentException();
+        }
+    }
+
+    public SortedMap<K, V> headMap(K toKey) {
+        return new Submap(null, toKey);
+    }
+
+    public SortedMap<K, V> tailMap(K fromKey) {
+        return new Submap(fromKey, null);
+    }
+
+    // ========== Infrastructure ==========
+
+    static <K, V> Iterator<K> descendingKeyIterator(SortedMap<K, V> map) {
+        return new Iterator<>() {
```

Comment:
Probably create `descendingEntryIterator` as base, and derive `descendingKeyIterator` and `descendingValueIterator`? This way, `map.get()` inside `descendingValueIterator` and `descendingEntryIterator` could be avoided.

Reply:
Not sure this will help. The fundamentals of advancing the descending iterator for SortedMap involve getting the last key of the current view (via the `lastKey` method) and then excluding it from the view with `headMap(lastKey)`. This allows iterating the keys in reverse without any `get` calls (though of course the inclusive/exclusive searching for the last key does involve probing the map). To build a descending entrySet iterator, one would have to do all of the above and then add a `get` call to obtain the value to construct the Entry. This adds a `get` call here but then saves a `get` call elsewhere. Seems like a wash to me.

It's kind of a shame because `TreeMap` internally has `getLastEntry`. Its `lastKey` method calls it and then throws away the Entry. :-(